source/_layouts/has_complimentary.liquid

```diff
@@ -0,0 +1,34 @@
+---
+layout: base
+---
+
+<div class="sl-l-large-holy-grail">
+  <div class="sl-l-large-holy-grail__body">
+    <div class="sl-l-large-holy-grail__main">
+      <div class="sl-l-container sl-l-container--small">
+        {% if before_introduction_include %}
+          {% include before_introduction_include %}
+        {% endif %}
+        <div class="sl-c-introduction">
+          {{ introduction }}
+        </div>
+
+        {{ content | typogr }}
+      </div>
+    </div>
+
+    {% if complementary_content %}
+      <div class="sl-l-large-holy-grail__complementary">
+        {% if complementary_content_nav %}
+          <nav aria-label={{ complementary_content_aria_label }} class="sl-c-list-navigation-wrapper">
+          {% endif %}
+          {% markdown %}
+            {{ complementary_content }}
+          {% endmarkdown %}
+          {% if complementary_content_nav %}
+          </nav>
+        {% endif %}
```

Comment:
Hrm, we're consistently running into this issue, which is basically "How to replicate [Nunjucks blocks](https://mozilla.github.io/nunjucks/templating.html#block) (that is, multiple named content 'slots') in Liquid?"

I haven't found a good solution to that -- see one attempt in `has_no_sidebars.liquid` using a dynamic include (`before_introduction_include`). That works, but means a separate included template for each different usage, which isn't ideal -- it seems the contents should live with the relevant template file.

@mirisuzanne Do you have an idea here? Or maybe we should check with Zach? It's frustrating enough at this point that I'm almost tempted to switch back to Nunjucks templates...

Reply:
Yea a separate include template for each use case isn't ideal 😕 I'd vote for reconsidering Liquid if we're on the fence about it. This is my first time using it and I haven't found its docs to be very friendly, especially compared to Nunjucks.

Reply:
I'd be fine reverting to Nunjucks if we're not feeling great about Liquid. On our other projects we often solve this with a macro, but Liquid doesn't have those either? I suppose a paired shortcode would provide the same thing. Really it's a matter of capturing the contents in a variable/argument, ouside the logic - and then using if/else to render that variable/arg with or without the wrapper.

Reply:
@sanajaved7 @mirisuzanne Before we give up on Liquid I think we should check with Zach to see if there's a better way to do this. I agree that Nunjucks is more powerful, but I'm also worried about its maintenance/lifecycle at this point. 🤷 

In the meantime, I realized that front-matter is more powerful than I thought, especially when using `eleventyComputed` so we can take advantage of filters/variables. (Even without that, it can include plain HTML.)

So I pushed a proposed solution in: https://github.com/oddbird/sass-site/commit/a730f6bdbb638990191146253838e39d861ac382

What do you think?

Reply:
@jgerigmeyer this looks good! 🎉 I was actually just reading more about Eleventy's front matter [options](https://www.11ty.dev/docs/data-frontmatter/) because you can specify different formats. This [Javascript format](https://www.11ty.dev/docs/data-frontmatter/#javascript-front-matter) is particularly interesting but it only works with Nunjucks / not Liquid.

Reply:
> This [Javascript format](https://www.11ty.dev/docs/data-frontmatter/#javascript-front-matter) is particularly interesting but it only works with Nunjucks / not Liquid.

Yep, agreed. Although it might be cleaner to use filters and yaml format regardless -- I'm not sure inlining JS in a template seems like a great pattern unless it's really a one-off case.

Reply:
@mirisuzanne What do you think of (current) proposed solution, vs checking with Zach on the best approach or switching to Nunjucks?

Reply:
The proposed solution just moves the nav into the content? I don't love that unless it's really a one-off? Otherwise, I would go for something with a little repetition - but simplified:

```liquid
{% if complementary_content_nav %}
  <nav aria-label={{ complementary_content_aria_label }} class="sl-c-list-navigation-wrapper">
    {{ complementary_content | markdown }}
   </nav>
{% else %}
  {{ complementary_content | markdown }}
{% endif %}
```

FWIW, I followed up on the general liquid recommendation, and Zach's preference is mostly specific to async rendering:

> Liquid is faster and async through-and-through. Nunjucks has some weird edge cases when you cross async/sync boundaries, perhaps most obvious with the set and setAsync shortcodes.

Reply:
> The proposed solution just moves the nav into the content? I don't love that unless it's really a one-off? 

Hm, that's not how I would describe it. The nav is *part* of the content IMO -- Sana abstracted it so that it could be used in one context but not used otherwise. In the old code, there's a single `complementary` block that can be filled with arbitrary contents -- in one or two places the contents includes some navigation (but with different classes and navs -- while in other places it doesn't. So I think the nav part of it is secondary to the bigger question of how to handle multiple content blocks in a layout, filled from a given template.

> FWIW, I followed up on the general liquid recommendation, and Zach's preference is mostly specific to async rendering:
> 
> > Liquid is faster and async through-and-through. Nunjucks has some weird edge cases when you cross async/sync boundaries, perhaps most obvious with the set and setAsync shortcodes.

Interesting, okay. Liquid had a release last month -- the most recent Nunjucks release was last June. Maybe I'm placing too much emphasis on support/maintenance tho? I don't think speed is a big priority for me in a static-site generator... 🤔  I guess we either keep trying to find solutions that work well enough with Liquid, or we ditch it for Nunjucks or another templating language that has more power/flexibility?

Reply:
I'm looking again at the options for how to handle the "nav" tag question -- I like them both better than what I had originally and I think I prefer Miriam's solution a bit more because I don't think I would expect the nav tag to be part of the content in the front matter. Tags feel like they belong more with the template itself rather than with the content. To keep it within the content, the nav tag feels a bit more hidden.